dali/operators/random/uniform_distribution.h

```diff
@@ -0,0 +1,173 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_OPERATORS_RANDOM_UNIFORM_DISTRIBUTION_H_
+#define DALI_OPERATORS_RANDOM_UNIFORM_DISTRIBUTION_H_
+
+#include <vector>
+#include "dali/operators/random/rng_base.h"
+#include "dali/pipeline/operator/arg_helper.h"
+#include "dali/core/dev_buffer.h"
+
+#define DALI_UNIFORM_DIST_TYPES (uint8_t, int8_t, uint16_t, int16_t, uint32_t, int32_t, uint64_t, \
+                                 int64_t, float16, float, double)
+
+namespace dali {
+
+template <typename T>
+class uniform_int_values_dist {
+ public:
+  uniform_int_values_dist() : values_(nullptr), nvalues_(0) {}
+  uniform_int_values_dist(const T *values, int64_t nvalues)
+    : values_(values), nvalues_(nvalues), dist_(0, nvalues-1) {}
+
+  template <typename Generator>
+  inline T operator()(Generator& g) {
+    int idx = dist_(g);
+    return values_[idx];
+  }
+
+ private:
+  const T *values_ = nullptr;
+  int64_t nvalues_ = 0;
+  std::uniform_int_distribution<int> dist_;
+};
+
+template <typename Backend>
+class UniformDistribution : public RNGBase<Backend, UniformDistribution<Backend>> {
+ public:
+  template <typename T>
+  struct DistContinuous {
+    using FloatType =
+      typename std::conditional<
+          ((std::is_integral<T>::value && sizeof(T) >= 4) || sizeof(T) > 4),
+          double, float>::type;
+
+    using type =
+      typename std::conditional_t<std::is_same<Backend, GPUBackend>::value,
+          curand_uniform_dist<FloatType>,
+          std::uniform_real_distribution<FloatType>>;
+  };
+
+  template <typename T>
+  struct DistDiscrete {
+    // TODO(janton): work directly with T values? (It'll need converting the values)
+    using type = typename std::conditional_t<std::is_same<Backend, GPUBackend>::value,
+        curand_uniform_int_values_dist<float>,
+        uniform_int_values_dist<float>>;
+  };
```

Comment:
So, let's say if `T == long long`. The `DistContinuous::FloatType` will be `double`, but for `DistDiscrete` we still have `float`. Will there be a loss of precision? If not, why the `double`?

On top of that, why the `typename T`? This struct doesn't seem to be using `T` at all...

Reply:
The values are provided to the operator as `float`, it wouldn't matter if we use anything different here, since the values would have to be converted anyway.

The reason why there is a `typename T` here is because in other distributions or even in this distribution but in the continuous case, the type is relevant.